rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -901,69 +921,129 @@ void PlayerImpl::load_storage_content()
+      {
+        // Stop sleeping if cancelled
+        if (std::atomic_exchange(&cancel_wait_for_next_message_, false)) {
+          break;
+        }
       }
-    }
-    std::lock_guard<std::mutex> lk(skip_message_in_main_play_loop_mutex_);
-    if (rclcpp::ok()) {
-      if (skip_message_in_main_play_loop_) {
-        skip_message_in_main_play_loop_ = false;
-        cancel_wait_for_next_message_ = false;
-        message_ptr = peek_next_message_from_queue();
-        continue;
+
+      std::lock_guard<std::mutex> lk_skip_message(skip_message_in_main_play_loop_mutex_);
+      if (rclcpp::ok()) {
+        // This means that the message we took from the queue's top was invalidated after a seek(),
+        // so we need to take a fresh element from the top of the queue, unless we have to stop
+        if (skip_message_in_main_play_loop_) {
+          skip_message_in_main_play_loop_ = false;
+          cancel_wait_for_next_message_ = false;
+          if (stop_playback_.load()) {
+            break;
+          } else {
+            message_ptr = take_next_message_from_queue();
+            continue;
+          }
+        }
+
+        const bool message_published = publish_message(message_ptr);
+        // If we tried to publish because of play_next(), jump the clock
+        if (play_next_.load()) {
+          clock_->jump(message_ptr->recv_timestamp);
+          // If we successfully played next, notify that we're done, otherwise keep trying
+          if (message_published) {
+            play_next_ = false;
+
+            std::lock_guard<std::mutex> lk(finished_play_next_mutex_);
+            finished_play_next_ = true;
+            play_next_result_ = true;
+            finished_play_next_cv_.notify_all();
+          }
+        }
       }
-      publish_message(message_ptr);
+      message_ptr = take_next_message_from_queue();
+    }
+
+    // At this point, we're at the end of the playback round, there are no more messages to play
+    // If we're still trying to play next or just got a request, we did not succeed
+    if (play_next_.exchange(false)) {
+      std::lock_guard<std::mutex> lk(finished_play_next_mutex_);
+      finished_play_next_ = true;
+      play_next_result_ = false;
+      finished_play_next_cv_.notify_all();
+    }
+
+    // While we're in a pause state, make sure we don't return if we happen to be at the end of the
+    // queue or playback round. However, if we get a request for play next during sleep_until(...),
+    // we need to stop waiting here and proceed to handle the play next request by doing another
+    // loop of this while().
+    while (!stop_playback_ && is_paused() && !play_next_.load() && rclcpp::ok()) {
+      clock_->sleep_until(clock_->now());
+    }
+    // If we ran out of messages and are not in pause state, it means we're done playing,
+    // unless play_next() is resuming and pausing the clock in order to wake us up
+    if (!is_paused() && !play_next_.load()) {
+      break;
+    }
```

Comment:
As far as I remember we have guards and situation when `play_next == true` and `is_paused == false` is not possible.
Then it seems we can simplify logic and only check.
```cpp
    if (!is_paused()) {
      break;
    }
```

Reply:
we cannot, because of what the comment is saying:

> unless play_next() is resuming and pausing the clock in order to wake us up

There's a race:

1. `play_messages_from_queue()` waits in this `while` if it is paused (which just means that the clock is paused) and there is no play next request:
    ```cpp
    while (!stop_playback_ && is_paused() && !play_next_.load() && rclcpp::ok()) {
      clock_->sleep_until(clock_->now());
    }
    ```
2. `play_next()` makes a request:
    ```cpp
     play_next_ = true;
    ```
3. `play_messages_from_queue()` exits from the `while` from (1)
4. `play_next()` wakes up `play_messages_from_queue()` from `sleep_until()` by resuming the clock:
    ```cpp
    // Temporarily resume clock to force wakeup in clock_->sleep_until(time),
    // then return in pause mode to preserve original state of the player
    clock_->resume();
    ```
5. `play_messages_from_queue()` then `break`s out of the outer `while` because the clock is now paused
    ```cpp
    // If we ran out of messages and are not in pause state, it means we're done playing,
    // unless play_next() is resuming and pausing the clock in order to wake us up
    if (!is_paused()) { // && !play_next_.load()) {
      break;
    }
    ```
5. `play_next()` then pauses the clock again, since its goal was just to stop `sleep_until()` by resuming the clock:
    ```cpp
    clock_->pause();
    ```
6. Now `play_messages_from_queue()` has exited, and no play next request will succeed again

I was able to reproduce this a few times in tests. The solution was to add `!play_next_.load()` to the `if` from (5). However, after merging #1869, we can indeed remove `!play_next_.load()` because we won't be using `resume()`+`pause()` to wake up.

Reply:
Okay. Thanks for the detailed explanation. I was misled about `is_paused()`. I was under the impression that we rely on some private member variable for pause detection. But we are actually relying on the underlying `clock_` state, which we resume temporarily.